build

```diff
@@ -0,0 +1,178 @@
+#!/usr/bin/env bash
+
+# build - build Jellyfin binaries or packages
+
+set -o errexit
+set -o pipefail
+
+# The list of possible package actions (except 'clean')
+actions=( 'build' 'package' 'sign' 'publish' )
+
+# The list of possible platforms, based on directories under 'deployment/'
+platforms=( $(
+    find deployment/ -maxdepth 1 -mindepth 1 -type d | sed 's/deployment\///' | sort
+) )
+
+# The list of standard dependencies required by all build scripts; individual
+# action scripts may specify their own dependencies
+dependencies=( 'tar' 'zip' )
+
+usage() {
+    echo -e "build - build Jellyfin binaries or packages"
+    echo -e ""
+    echo -e "Usage:"
+    echo -e " $ build --list-platforms"
+    echo -e " $ build --list-actions <platform>"
+    echo -e " $ build [-b/--web-branch <web_branch>] <platform> <action>"
+    echo -e ""
+    echo -e "The web_branch defaults to 'master'."
+    echo -e "To build all platforms, use 'all'."
+    echo -e "To perform all build actions, use 'all'."
+    echo -e "Build output files are collected at '../jellyfin-build/<platform>'."
+}
+
+if [[ -z $1 ]]; then
+    usage
+    exit 1
+fi
+
+# List all available platforms then exit
+if [[ $1 == '--list-platforms' ]]; then
+    echo -e "Available platforms:"
+    for platform in ${platforms[@]}; do
+        echo -e "  ${platform}"
+    done
+    exit 0
+fi
+
+# List all available actions for a given platform then exit
+if [[ $1 == '--list-actions' ]]; then
+    platform="$2"
+    if [[ ! " ${platforms[@]} " =~ " ${platform} " ]]; then
+        echo "Platform $platform does not exist."
+        exit 1
+    fi
+    echo -e "Available actions for platform ${platform}:"
+    for action in ${actions[@]}; do
+        if [[ -f deployment/${platform}/${action}.sh ]]; then
+            echo -e "  ${action}"
+        fi
+    done
+    exit 0
+fi
+
+# Parse branch option
+if [[ $1 == '-b' || $1 == '--web-branch' ]]; then
+	web_branch="$2"
+	shift 2
+else
+    web_branch="master"
+fi
+
+# Parse platform option
+if [[ -n $1 ]]; then
+    platform=( "$1" )
+    shift
+fi
+if [[ $platform == 'all' ]]; then
+    platform=( ${platforms[@]} )
+else
+    if [[ ! " ${platforms[@]} " =~ " ${platform} " ]]; then
+        echo "Platform $platform is invalid. Use the '--list-platforms' option to list available platforms."
+        exit 1
+    fi
+fi
+
+# Parse action option
+if [[ -n $1 ]]; then
+    action=( "$1" )
+    shift
+fi
+if [[ $action == 'all' ]]; then
+    action=( ${actions[@]} )
+else
+    if [[ ! " ${actions[@]} " =~ " ${action} " ]]; then
+        echo "Action $action is invalid. Use the '--list-actions <platform>' option to list available actions."
+        exit 1
+    fi
+fi
+
+# Verify required utilities are installed
+missing_deps="no"
+for utility in ${dependencies[@]}; do
+    if ! which ${utility} &>/dev/null; then
+        missing_deps="yes"
```

Comment:
Well this is the "main list", things that the script shouldn't even try to run without. It'll probably be these two things (plus `dotnet-sdk` which I totally forgot!) only for the near future.

I also added support for individual platforms setting their own dependency lists (see the next block down).

The main issue I have with trying to determine "what the user has installed" is the myriad of package managers. It's far far easier and cleaner from the script side to just dump the list and have the user determine which ones they're missing. I don't expect these lists to be more than a few entries each, otherwise it should probably go in Docker ;-)

Reply:
OH I see what you mean! I was misinterpreting. I can make it do that yes.